src/nvim/eval/vars.c

```diff
@@ -1900,28 +1904,45 @@ static OptVal tv_to_optval(typval_T *tv, const char *option, uint32_t flags, boo
 
 /// Convert an option value to typval.
 ///
-/// @param[in]  value  Option value to convert.
+/// @param[in]  value    Option value to convert.
+/// @param      numbool  Whether to convert boolean values to number.
+///                      Used for backwards compatibility.
 ///
 /// @return  OptVal converted to typval.
-typval_T optval_as_tv(OptVal value)
+typval_T optval_as_tv(OptVal value, bool numbool)
 {
   typval_T rettv = { .v_type = VAR_SPECIAL, .vval = { .v_special = kSpecialVarNull } };
 
   switch (value.type) {
   case kOptValTypeNil:
     break;
   case kOptValTypeBoolean:
-    switch (value.data.boolean) {
-    case kTrue:
-      rettv.v_type = VAR_BOOL;
-      rettv.vval.v_bool = kBoolVarTrue;
-      break;
-    case kFalse:
-      rettv.v_type = VAR_BOOL;
-      rettv.vval.v_bool = kBoolVarFalse;
-      break;
-    case kNone:
-      break;  // return v:null for None boolean value
+    if (numbool) {
+      switch (value.data.boolean) {
+      case kTrue:
+        rettv.v_type = VAR_NUMBER;
+        rettv.vval.v_number = 1;
+        break;
+      case kFalse:
+        rettv.v_type = VAR_NUMBER;
+        rettv.vval.v_number = 0;
+        break;
+      case kNone:
+        break;  // return v:null for None boolean value
+      }
+    } else {
+      switch (value.data.boolean) {
+      case kTrue:
+        rettv.v_type = VAR_BOOL;
+        rettv.vval.v_bool = kBoolVarTrue;
+        break;
+      case kFalse:
+        rettv.v_type = VAR_BOOL;
+        rettv.vval.v_bool = kBoolVarFalse;
+        break;
+      case kNone:
+        break;  // return v:null for None boolean value
```

Comment:
could dedupe this a bit.

```suggestion
    rettv.v_type = numbool ? VAR_BOOL : numbool;
    switch (value.data.boolean) {
    case kTrue:
      rettv.vval.v_number = 1;
      rettv.vval.v_bool = kBoolVarTrue;
      break;
    case kFalse:
      rettv.vval.v_number = 0;
      rettv.vval.v_bool = kBoolVarFalse;
      break;
    case kNone:
      break;  // return v:null for None boolean value
    }
```

Reply:
> could dedupe this a bit.

Thanks for this, but setting both `v_number` and `v_bool` would probably not work as intended as `vval` is a union, not a struct. But surrounded that part by an `if` statement could work

Reply:
```c
  case kOptValTypeBoolean:
    if (value.data.boolean != kNone) {
      if (numbool) {
        rettv.v_type = VAR_NUMBER;
        rettv.vval.v_number = value.data.boolean == kTrue;
      } else {
        rettv.v_type = VAR_BOOL;
        rettv.vval.v_bool = value.data.boolean == kTrue;
      }
    }
    break;  // return v:null for None boolean value.
```

This seems to work